src/vs/editor/contrib/stickyScroll/browser/stickyScrollProvider.ts

```diff
@@ -0,0 +1,172 @@
+		@ILanguageFeaturesService _languageFeaturesService: ILanguageFeaturesService,
+	) {
+		super();
+		this.editor = editor;
+		this.languageFeaturesService = _languageFeaturesService;
+		this.updateSoon = this._register(new RunOnceScheduler(() => this.update(), 50));
+		this._register(this.editor.onDidChangeConfiguration(e => {
+			if (e.hasChanged(EditorOption.experimental)) {
+				this.readConfiguration();
+			}
+		}));
+		this.readConfiguration();
+	}
+
+	private readConfiguration() {
+		const options = this.editor.getOption(EditorOption.experimental);
+		if (options.stickyScroll.enabled === false) {
+			this.sessionStore.clear();
+			return;
+		} else {
+			this.sessionStore.add(this.editor.onDidChangeModel(() => this.update()));
+			this.sessionStore.add(this.editor.onDidChangeHiddenAreas(() => this.update()));
+			this.sessionStore.add(this.editor.onDidChangeModelContent(() => this.updateSoon.schedule()));
+			this.sessionStore.add(this.languageFeaturesService.documentSymbolProvider.onDidChange(() => this.update()));
+			this.update();
+		}
+	}
+
+	public getVersionId() {
+		return this.modelVersionId;
+	}
+
+	private async update(): Promise<void> {
+		this.cts?.dispose(true);
+		this.cts = new CancellationTokenSource();
+		await this.updateOutlineModel(this.cts.token);
+		this.onStickyScrollChangeEmitter.fire();
+	}
+
+	private async updateOutlineModel(token: CancellationToken) {
+		if (this.editor.hasModel()) {
+			const model = this.editor.getModel();
+			const modelVersionId = model.getVersionId();
+			const outlineModel = await OutlineModel.create(this.languageFeaturesService.documentSymbolProvider, model, token) as OutlineModel;
+			if (token.isCancellationRequested) {
+				return;
+			}
+			this.outlineModel = StickyOutlineElement.fromOutlineModel(outlineModel);
+			this.modelVersionId = modelVersionId;
+		}
+	}
+
+	public getCandidateStickyLinesIntersectingFromOutline(range: StickyRange, outlineModel: StickyOutlineElement, result: StickyLineCandidate[], depth: number, lastStartLineNumber: number): void {
+		let lastLine = lastStartLineNumber;
+		for (const child of outlineModel.children) {
+			if (child.range) {
+				const childStartLine = child.range.startLineNumber;
+				const childEndLine = child.range.endLineNumber;
+				if (range.startLineNumber <= childEndLine + 1 && childStartLine - 1 <= range.endLineNumber && childStartLine !== lastLine) {
+					lastLine = childStartLine;
+					result.push(new StickyLineCandidate(childStartLine, childEndLine - 1, depth + 1));
+					this.getCandidateStickyLinesIntersectingFromOutline(range, child, result, depth + 1, childStartLine);
+				}
+			} else {
+				this.getCandidateStickyLinesIntersectingFromOutline(range, child, result, depth, lastStartLineNumber);
+			}
+		}
+	}
+
+	public getCandidateStickyLinesIntersecting(range: StickyRange): StickyLineCandidate[] {
+		let stickyLineCandidates: StickyLineCandidate[] = [];
+		if (range.startLineNumber && range.endLineNumber) {
```

Comment:
When would this `if` statement be `false`?

Reply:
I previously noticed that at the bottom of the file, we can get the following error : Cannot read properties of undefined (reading 'startLineNumber'): TypeError: Cannot read properties of undefined (reading 'startLineNumber') on the function getVisibleRanges(). I think instead of doing a check here, I'll put a check at the moment when we call the getVisibleRanges method

Reply:
Do you mean to check if `getVisibleRanges()` returns an empty array?